### ISSUE TITLE
PMacc ForEach documentation

### ISSUE THREAD

**Issue by Heikman** [ComputationalRadiationPhysics/picongpu#842]
Please clarify documentation of algorithms::ForEach:

- https://github.com/ComputationalRadiationPhysics/picongpu/blob/dev/src/libPMacc/include/algorithms/ForEach.hpp#L196
 - *CallIonization* is a binary functor and it works too, although *ForEach* wants to have unary (lambda) functors.
 - the name "*lambda* functor" is misleading, because *lambda* means there is no name, but I guess you allow for ordinary named functors too. A *lambda expression/functor* is just like ``_1 + _2``
- https://github.com/ComputationalRadiationPhysics/picongpu/blob/fb2ca7e202a8d5f7a3d92fa56398130be6aa6143/src/libPMacc/include/algorithms/ForEach.hpp#L202-L203
 - can you please check the syntax. What does the double parentheses mean? Is the first one a compile-time argument? *Functor* looks like a binary operator. Is this why *CallIonization* is binary too?
- Nice to have: A compiling piece of example code.

**Comment by psychocoderHPC** [comment]
- Yes you are right `lambda` functor is not the right name for this, it should be `binary operator`
- [*CallIonization*](https://github.com/ComputationalRadiationPhysics/picongpu/blob/dev/src/picongpu/include/particles/ParticlesFunctors.hpp#L159) is a unary operator (it has only one template argument)

```
       call:   ForEach<MPLSeq,Functor,Accessor>()(42);
       unrolled code: Functor(Accessor(int))(42);
                               Functor(Accessor(float))(42);
```
You need to read the definition above: `definition: F(X) means boost::apply<F,X>`
Whis the definition in mind you got:
```
typedef typename boost::apply<Accessor,int>::type Type1;
typedef typename boost::apply<Functor,Type1>::type SolvedUnaryOperator;
SolvedUnaryOperator()(42);
```

A example where `ForEach` is used: https://github.com/ComputationalRadiationPhysics/picongpu/blob/dev/src/picongpu/include/simulationControl/MySimulation.hpp#L268-L269

I hope this helps you a little bit.

**Comment by psychocoderHPC** [comment]
In the `ForEach` example there are two parentheses missed:

`Functor(Accessor(int))(42);` should be `Functor(Accessor(int))()(42);`

**Comment by ax3l** [comment]
[Example code](https://github.com/ComputationalRadiationPhysics/picongpu/pull/790/files#r29332248) :)

**Comment by ax3l** [comment]
Pls let us update the findings from above with a pull request :sparkles:

**Comment by psychocoderHPC** [comment]
yes please do not close this issue befor the documentation is fixed

**Comment by ax3l** [comment]
-> transformed from question to task
